### PR TITLE
Mongo's `MongoQueryExecutor`

### LINKED ISSUE #1306: Add a MongoQueryExecutor interface that you can extend from
### Feature description

Similar to https://github.com/micronaut-projects/micronaut-data/blob/master/data-model/src/main/java/io/micronaut/data/repository/jpa/JpaSpecificationExecutor.java it should provide methods that take a MongoDB filter. See https://mongodb.github.io/mongo-java-driver/3.4/driver/getting-started/quick-start/#specify-a-query-filter

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes #1306

### CODE REVIEW

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    Optional<E> findOne(MongoFindOptions options);
+
+    /**
+     * Finds all results.
+     *
+     * @param filter The filter to be applied
+     * @return The records
+     */
+    List<E> findAll(@Nullable Bson filter);
```
[the remark this answers is not in the document]
Comment (graemerocher) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:59]: Wouldn't it be better to return a `FindIterable`? What about pagination?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Finds all results.
+     *
+     * @param options The options
+     * @return The records
+     */
+    List<E> findAll(MongoFindOptions options);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:67]: Is the argument `@Nullable`? Add `@NonNull` to return type. 
Wouldn't it be better to return a FindIterable? What about pagination?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Finds one result.
+     *
+     * @param pipeline The pipeline to be applied
+     * @return The optional result
+     */
+    Optional<E> findOne(Iterable<Bson> pipeline);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:75]: Is the argument `@Nullable`?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Finds one result.
+     *
+     * @param pipeline The pipeline to be applied
+     * @param options  The aggregation options
+     * @return The optional result
+     */
+    Optional<E> findOne(Iterable<Bson> pipeline, MongoAggregationOptions options);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:84]: are the arguments `@Nullable`?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Finds all results.
+     *
+     * @param pipeline The pipeline to be applied
+     * @return The results
+     */
+    List<E> findAll(Iterable<Bson> pipeline);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:92]: Is the argument `@Nullable`? Add `@NonNull` to return type. 
Wouldn't it be better to return a FindIterable? What about pagination?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Finds all results.
+     *
+     * @param pipeline The pipeline to be applied
+     * @param options  The options
+     * @return The results
+     */
+    List<E> findAll(Iterable<Bson> pipeline, MongoAggregationOptions options);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:101]: Is the argument `@Nullable`? Add `@NonNull` to return type. 
Wouldn't it be better to return a FindIterable? What about pagination?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    long count(@Nullable Bson filter);
+
+    /**
+     * Delete the records matching the filter.
+     *
+     * @param filter The filter to be applied
+     * @return The deleted count
+     */
+    long deleteAll(Bson filter);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:117]: Is the argument `@Nullable`?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Delete the records matching the filter.
+     *
+     * @param filter  The filter to be applied
+     * @param options The delete options
+     * @return The deleted count
+     */
+    long deleteAll(Bson filter, DeleteOptions options);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:126]: Is the argument `@Nullable`?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Update the records matching the filter.
+     *
+     * @param filter The filter to be applied
+     * @param update The update modification
+     * @return The updated count
+     */
+    long updateAll(Bson filter, Bson update);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:135]: Is the argument `@Nullable`?

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Update the records matching the filter.
+     *
+     * @param filter  The filter to be applied
+     * @param update  The update modification
+     * @param options The update options
+     * @return The updated count
+     */
+    long updateAll(Bson filter, Bson update, UpdateOptions options);
```
Comment (graemerocher) [inline data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:145]: Is the argument `@Nullable`?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dstepanov) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:59]: I have added nonnulls, `FindIterable` return types would require more work.

[the remark this answers is not in the document]
Comment (dstepanov) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:59]: Yeah pagination might be possible to add.

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java (hunk 5 above)
Comment (dstepanov) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java:92]: You can do everything else without `FindIterable`

Comment (graemerocher) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java]: I don't see a test for pagination?

**Comment by graemerocher** [comment]
and the coverage could be improved :)

**Comment by dstepanov** [comment]
It looks like annotation processing code is not being included in the coverage report, it's only included when run by tests.

File: data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java (hunk 5 above)
Comment (dstepanov) [reply data-mongodb/src/main/java/io/micronaut/data/mongodb/repository/MongoQueryExecutor.java]: Added pagination

**Comment by graemerocher** [comment]
Hmmm yeah

**Pull request merged**